### PR TITLE
Avoid redundant copying of tars during kube-up for gce if the same file already exists

### LINKED ISSUE #46791: Avoid copying tars from local machine to gce during kube-up if the same tars already exist
**Is this a request for help?**: No

**What keywords did you search in Kubernetes issues before filing this one?** : copy-to-staging

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Feature Request / Refactoring. 

**Environment**:
- **Cloud provider or hardware configuration**: Google Cloud (gce)

**What happened**:

Whenever I execute cluster/kube-up.sh it copies my tar files to google cloud, even if the files haven't changed. These files are a few hundred megs and can take a long time to upload.

**What you expected to happen**:

I'd like cluster/gce/utils.sh copy-to-staging function to check whether a matching tar already exist, and avoid uploading it again in this situation.

**How to reproduce it**:
execute cluster/kube-up.sh

Here is the current behavior:
cluster/kube-up.sh 
... Starting cluster in us-central1-b using provider gce
... calling verify-prereqs
... calling verify-kube-binaries
... calling kube-up
Project: <PROJECT-NAME>
Zone: us-central1-b
+++ Staging server tars to Google Storage: gs://kubernetes-staging-<PROJECT-NAME>/kubernetes-devel
+++ kubernetes-server-linux-amd64.tar.gz uploaded (sha1 = <HASH>)

**Anything else we need to know**:

I have written the code to deal with this issue, and I'll submit a related PR shortly.

### PR DESCRIPTION
**Pull request by ianchakeres**
**What this PR does / why we need it**: 

Whenever I execute cluster/kube-up.sh it copies my tar files to google cloud, even if the files haven't changed. This PR checks to see whether the files already exist, and avoids uploading them again. These files are large and can take a long time to upload.

**Which issue this PR fixes**: fixes #46791

**Special notes for your reviewer**:

Here is the new output:

cluster/kube-up.sh 
... Starting cluster in us-central1-b using provider gce
... calling verify-prereqs
... calling verify-kube-binaries
... calling kube-up
Project: PROJECT
Zone: us-central1-b
+++ Staging server tars to Google Storage: gs://kubernetes-staging-PROJECT/kubernetes-devel
+++ kubernetes-server-linux-amd64.tar.gz uploaded earlier, cloud and local file md5 match (md5 = 3a095kcf27267a71fe58f91f89fab1bc)

**Release note**:
```cluster/kube-up.sh on gce now avoids redundant copying of kubernetes tars if the local and cloud files' md5 hash match```

### CODE REVIEW

**Comment by ianchakeres** [comment]
@pwittrock - please take a quick peak at this PR and let me know if you recommend any changes.

**Comment by ianchakeres** [comment]
@k8s-bot ok to test

**Comment by pwittrock** [comment]
@k8s-bot ok to test

**Comment by pwittrock** [comment]
@ianchakeres Please add a release note in the form show in the template:

e.g.
*Look at the markdown and copy the 3x-tick tags*:

```release-note
cluster/kube-up.sh now reuses tars it has already uploaded if the hash matches the local copy
```

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -154,11 +154,23 @@ function copy-to-staging() {
...
   local -r hash=$4
+  local -r basename_tar=$(basename ${tar})
+
+  #check whether this tar alread exists and has the same hash
+  #if it matches, then don't bother uploading it again
+  local -r remote_tar_hash=$(gsutil hash -h -m ${staging_path}/${basename_tar} 2>/dev/null | grep "Hash (md5):" | awk -F ':' '{print $2}')
+  if [[ -n ${remote_tar_hash} ]]; then
+    local -r local_tar_hash=$(gsutil hash -h -m ${tar} 2>/dev/null | grep "Hash (md5):" | awk -F ':' '{print $2}')
+    if [[ "${remote_tar_hash}" == "${local_tar_hash}" ]]; then
+      echo "+++ ${basename_tar} uploaded earlier and hash matches"
```
Comment (pwittrock) [inline cluster/gce/util.sh]: Maybe output the hash value as well

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -154,11 +154,23 @@ function copy-to-staging() {
   local -r gs_url=$2
   local -r tar=$3
   local -r hash=$4
+  local -r basename_tar=$(basename ${tar})
+
+  #check whether this tar alread exists and has the same hash
+  #if it matches, then don't bother uploading it again
+  local -r remote_tar_hash=$(gsutil hash -h -m ${staging_path}/${basename_tar} 2>/dev/null | grep "Hash (md5):" | awk -F ':' '{print $2}')
+  if [[ -n ${remote_tar_hash} ]]; then
+    local -r local_tar_hash=$(gsutil hash -h -m ${tar} 2>/dev/null | grep "Hash (md5):" | awk -F ':' '{print $2}')
```
Comment (pwittrock) [inline cluster/gce/util.sh]: this is fine, but would be a bit cleaner if pulled into a function to return the hash for a given name

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -154,11 +154,23 @@ function copy-to-staging() {
   local -r gs_url=$2
   local -r tar=$3
   local -r hash=$4
+  local -r basename_tar=$(basename ${tar})
+
+  #check whether this tar alread exists and has the same hash
+  #if it matches, then don't bother uploading it again
+  local -r remote_tar_hash=$(gsutil hash -h -m ${staging_path}/${basename_tar} 2>/dev/null | grep "Hash (md5):" | awk -F ':' '{print $2}')
```
Comment (pwittrock) [inline cluster/gce/util.sh]: How does this work against a remote tar?  I am not familiar with gsutil but the description is ` The hash command calculates hashes on a local file`.  Is this working on a remote file or a local file?

Comment (ianchakeres) [inline cluster/gce/util.sh]: In this particular case the hash is operating on the remote file, and I tested that it works.

Looks like that documentation you referenced is out of date, and this feature was added to support hash on cloud files in gsutil v4.21. Here is the related issue for tracking hash of cloud files - https://github.com/GoogleCloudPlatform/gsutil/issues/369

File: cluster/gce/util.sh (hunk 2 above)
Comment (ianchakeres) [inline cluster/gce/util.sh]: Ok. I'll update the next revision to use a separate function for calling gsutil to get the hash and parse the output.

File: cluster/gce/util.sh (hunk 1 above)
Comment (ianchakeres) [inline cluster/gce/util.sh]: Sounds good. Working on an update now.

**Comment by pwittrock** [comment]
@vishh Would you take a look at this since it looks like you can approve?

**Comment by ianchakeres** [comment]
@k8s-bot  pull-kubernetes-e2e-gce-etcd3 test this

**Comment by ianchakeres** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by ianchakeres** [comment]
Yesterday I encountered some test flakes inside e2e-gce-etcd3 on this PR #46792. I think they relate to these two reported flaky tests issues #43520 & #47446.

From my investigation the failures mention timeouts.

Here are the the failed test runs' output: 
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46792/pull-kubernetes-e2e-gce-etcd3/36002/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46792/pull-kubernetes-e2e-gce-etcd3/35975/
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46792/pull-kubernetes-e2e-gce-etcd3/35842/

**Comment by ianchakeres** [comment]
@vishh can you please take a quick look and approve this PR? Change is about ~25 lines and half are clarifying comments.

If you have any questions or comments, just let me know.

File: cluster/gce/util.sh [hunk 4]
```diff
@@ -148,19 +148,50 @@ function detect-project() {
...
+
 # Copy a release tar and its accompanying hash.
 function copy-to-staging() {
   local -r staging_path=$1
   local -r gs_url=$2
   local -r tar=$3
   local -r hash=$4
+  local -r basename_tar=$(basename ${tar})
+
+  #check whether this tar alread exists and has the same hash
```
Comment (jeffvance) [inline cluster/gce/util.sh]: typo: already

**Pull request merged**